File: OrangeJuice.Server.Core/Services/AwsProductProvider.cs
using System.Linq;
using System.Threading.Tasks;
using System.Xml.Linq;

using Ab.Amazon;
using Ab.Amazon.Data;
using Ab.Factory;

using OrangeJuice.Server.Data.Models;

namespace OrangeJuice.Server.Services
{
	public sealed class AwsProductProvider : IAwsProductProvider
	{
		private readonly IAwsClient _client;
		private readonly IFactory<ProductDescriptor, XElement, AwsProductSearchCriteria> _factory;

		public AwsProductProvider(IAwsClient client, IFactory<ProductDescriptor, XElement, AwsProductSearchCriteria> factory)
		{
			_client = client;
			_factory = factory;
		}

		public async Task<ProductDescriptor[]> Search(string barcode, BarcodeType barcodeType)
		{
			var searchCriteria = new AwsProductSearchCriteria
			{
				Operation = "ItemLookup",
				SearchIndex = "Grocery",
				ResponseGroups = new[] { "Images", "ItemAttributes" },
				IdType = barcodeType.ToString(),
				ItemId = barcode
			};

			var items = await _client.GetItems(searchCriteria);
			return items.Select(i => _factory.Create(i, searchCriteria)).ToArray();
		}
	}
}

File: OrangeJuice.Server.Core.FSharp/Services/AwsProductProvider.fs
namespace OrangeJuice.Server.FSharp.Services

open System.Threading.Tasks
open System.Xml.Linq

open Factory

open OrangeJuice.Server.Data.Models
open OrangeJuice.Server.Services

type AwsProductProvider(client : IAwsClient, factory : IFactory<ProductDescriptor, XElement, AwsProductSearchCriteria>) =
    interface IAwsProductProvider with
        member this.Search(barcode : string, barcodeType : BarcodeType) : Task<ProductDescriptor[]> =
            this.Search(barcode, barcodeType) |> Async.StartAsTask

    member this.Search(barcode : string, barcodeType : BarcodeType) = async {
        let searchCriteria = new AwsProductSearchCriteria(
            Operation = "ItemLookup",
            SearchIndex = "Grocery",
            ResponseGroups = [| "Images"; "ItemAttributes" |],
            IdType = barcodeType.ToString(),
            ItemId = barcode)

        let! items = client.GetItems(searchCriteria) |> Async.AwaitTask
        let seq = Seq.map (fun item -> factory.Create(item, searchCriteria)) items
        return seq |> Array.ofSeq
    }